src/main/java/com/linecorp/armeria/server/ServerInitializer.java

```diff
@@ -213,4 +200,62 @@ private void addHttpHandlers(ChannelHandlerContext ctx) {
             p.addLast(new HttpServerHandler(config, SessionProtocol.H1));
         }
     }
+
+    private final class Http2PrefaceOrHttpHandler extends ByteToMessageDecoder {
+
+        private String name;
+
+        @Override
+        public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+            super.handlerAdded(ctx);
+            name = ctx.name();
+        }
+
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+            if (in.readableBytes() < 4) {
+                return;
+            }
+
+            if (in.getInt(in.readerIndex()) == 0x50524920) { // If starts with 'PRI '
+                configureHttp2(ctx);
+            } else {
+                configureHttp1(ctx);
```

Comment:
configureHttp1WithUpgrade or configureHttpUpgrade? This codepath may still become http2 right?